cub/cub/device/dispatch/dispatch_transform.cuh

```diff
@@ -0,0 +1,1161 @@
+constexpr auto loaded_bytes_per_iteration() -> int
+{
+  return (int{sizeof(value_t<Its>)} + ... + 0);
+}
+#else // ^^^ C++17 ^^^ / vvv C++11 vvv
+constexpr int sum()
+{
+  return 0;
+}
+
+template <typename... Ts>
+constexpr int sum(int head, Ts... tail)
+{
+  return head + sum(tail...);
+}
+
+template <typename... Its>
+constexpr auto loaded_bytes_per_iteration() -> int
+{
+  return sum(int{sizeof(value_t<Its>)}...);
+}
+#endif // _CCCL_STD_VER >= 2017
+
+enum class Algorithm
+{
+  fallback_for,
+  prefetch,
+  unrolled_staged,
+  memcpy_async,
+#ifdef _CUB_HAS_TRANSFORM_UBLKCP
+  ublkcp,
+#endif // _CUB_HAS_TRANSFORM_UBLKCP
+};
+
+// TODO(bgruber): only needed so we can instantiate the kernel generically for any policy. Remove when fallback_for is
+// dropped.
+template <typename, typename Offset, typename F, typename RandomAccessIteartorOut, typename... RandomAccessIteartorIn>
+_CCCL_DEVICE void transform_kernel_impl(
+  ::cuda::std::integral_constant<Algorithm, Algorithm::fallback_for>,
+  Offset,
+  int,
+  F,
+  RandomAccessIteartorOut,
+  RandomAccessIteartorIn...)
+{}
+
+template <int BlockThreads>
+struct prefetch_policy_t
+{
+  static constexpr int BLOCK_THREADS = BlockThreads;
+  // items per tile are determined at runtime. these (inclusive) bounds allow overriding that value via a tuning policy
+  static constexpr int ITEMS_PER_THREAD_NO_INPUT = 2; // items per thread when no input streams exits (just filling)
+  static constexpr int MIN_ITEMS_PER_THREAD      = 1;
+  static constexpr int MAX_ITEMS_PER_THREAD      = 32;
+};
+
+// Prefetches (at least on Hopper) a 128 byte cache line. Prefetching out-of-bounds addresses has no side effects
+template <typename T>
+_CCCL_DEVICE _CCCL_FORCEINLINE void prefetch(const T* addr)
+{
+  assert(__isGlobal(addr));
+  // TODO(bgruber): prefetch to L1 may be even better
+  asm volatile("prefetch.global.L2 [%0];" : : "l"(addr) : "memory");
+}
+
+// TODO(bgruber): there is also the cp.async.bulk.prefetch instruction available on Hopper. May improve perf a tiny bit
+// as we need to create less instructions to prefetch the same amount of data.
+
+// overload for any iterator that is not a pointer, do nothing
+template <typename It, ::cuda::std::__enable_if_t<!::cuda::std::is_pointer<It>::value, int> = 0>
+_CCCL_DEVICE _CCCL_FORCEINLINE void prefetch(It)
+{}
+
+// this kernel guarantees stable addresses for the parameters of the user provided function
+template <typename PrefetchPolicy,
+          typename Offset,
+          typename F,
+          typename RandomAccessIteartorOut,
+          typename... RandomAccessIteartorIn>
+_CCCL_DEVICE void transform_kernel_impl(
+  ::cuda::std::integral_constant<Algorithm, Algorithm::prefetch>,
+  Offset num_items,
+  int num_elem_per_thread,
+  F f,
+  RandomAccessIteartorOut out,
+  RandomAccessIteartorIn... ins)
+{
+  constexpr int block_dim = PrefetchPolicy::BLOCK_THREADS;
+  {
+    const int tile_stride = block_dim * num_elem_per_thread;
```

Comment:
Question: do we launch the kernels with the right grid size? the problem is when `block_dim * num_elem_per_thread` is larger than `INT_MAX`